### ISSUE TITLE
Xml Validation with Xsd has failed: Attribute 'status' must appear on element 'step'

### ISSUE THREAD

**Issue by alb-i986** [HPSoftware/octane-cucumber-jvm#6]
When I run a pipeline, I can't see the test results on Octane.
It looks like they are not uploaded correctly, but actually it turns out they are, but the parsing of the XML file (OctaneGherkinResults) fails:

```
2017-08-22T13:22:05,836 | ERROR | CID:  | Global_Time: 2017-08-22T13:22:05.817 | Service: NGA | Thread: TestResultRequestExecutorImpl thread-4 | System_IP: 172.18.0.6 | Session:  | User:  | SSID: 1001 | WSID: 1001 | Class: rest.TestResultRequestExecutorImpl | pushXml: failed
com.hp.mqm.testbox.exceptions.TestResultFormatException: Xml Validation with Xsd has failed
In some cases this is caused from an inconsistency between the server version and either the Jenkins plugin and\or the Octane Cucumber Runner
It is highly recommended that you check the versions you are using and have installed
[ line: 97, column: 94 ] cvc-complex-type.4: Attribute 'status' must appear on element 'step'.
	at com.hp.mqm.testbox.TestResultPushServiceImpl.pushResults(TestResultPushServiceImpl.java:118) ~[test-box-service-12.55.5.100.jar:?]
	at com.hp.mqm.testbox.rest.TestResultAbstractPushResource.lambda$pushXml$1(TestResultAbstractPushResource.java:34) ~[test-box-rest-12.55.5.100.jar:?]
	at com.hp.mqm.testbox.rest.TestResultRequestExecutorImpl$MqmContextExecutor.lambda$null$6(TestResultRequestExecutorImpl.java:104) ~[test-box-rest-12.55.5.100.jar:?]
	at com.hp.mqm.ps.platformservices.context.GenericContextUtils.workWithContext(GenericContextUtils.java:329) ~[mqm-ps-platform-services-api-12.55.5.100.jar:?]
	at com.hp.mqm.ps.platformservices.context.GenericContextUtils.workWithContext(GenericContextUtils.java:309) ~[mqm-ps-platform-services-api-12.55.5.100.jar:?]
	at com.hp.mqm.ps.platformservices.context.GenericContextUtils.workWithWorkspace(GenericContextUtils.java:113) ~[mqm-ps-platform-services-api-12.55.5.100.jar:?]
	at com.hp.mqm.testbox.rest.TestResultRequestExecutorImpl$MqmContextExecutor.lambda$executeInContext$7(TestResultRequestExecutorImpl.java:103) ~[test-box-rest-12.55.5.100.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_71]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_71]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_71]
Caused by: org.owasp.esapi.errors.ValidationException: [ line: 97, column: 94 ] cvc-complex-type.4: Attribute 'status' must appear on element 'step'.
	at com.hp.mqm.platform.security.InputValidationServiceImpl.throwXMLValidationException(InputValidationServiceImpl.java:444) ~[mqm-infra-services-impl-12.55.5.100.jar:?]
	at com.hp.mqm.platform.security.InputValidationServiceImpl.validateSecureXML(InputValidationServiceImpl.java:404) ~[mqm-infra-services-impl-12.55.5.100.jar:?]
	at com.hp.mqm.testbox.TestResultPushServiceImpl.pushResults(TestResultPushServiceImpl.java:109) ~[test-box-service-12.55.5.100.jar:?]
	... 9 more
Caused by: org.xml.sax.SAXException: [ line: 97, column: 94 ] cvc-complex-type.4: Attribute 'status' must appear on element 'step'.
	at com.hp.mqm.platform.security.InputValidationServiceImpl$MyErrorHandler.rethrowWithCoordinates(InputValidationServiceImpl.java:542) ~[mqm-infra-services-impl-12.55.5.100.jar:?]
	at com.hp.mqm.platform.security.InputValidationServiceImpl$MyErrorHandler.error(InputValidationServiceImpl.java:533) ~[mqm-infra-services-impl-12.55.5.100.jar:?]
	at com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.error(ErrorHandlerWrapper.java:137) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:396) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:284) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator$XSIErrorReporter.reportError(XMLSchemaValidator.java:452) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.reportSchemaError(XMLSchemaValidator.java:3230) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.addDefaultAttributes(XMLSchemaValidator.java:2936) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.handleStartElement(XMLSchemaValidator.java:2054) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.startElement(XMLSchemaValidator.java:740) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.jaxp.validation.ValidatorHandlerImpl.startElement(ValidatorHandlerImpl.java:570) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.startElement(AbstractSAXParser.java:509) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.parsers.AbstractXMLDocumentParser.emptyElement(AbstractXMLDocumentParser.java:182) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:357) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2787) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:606) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:118) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:510) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:848) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:777) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1213) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.jaxp.validation.ValidatorHandlerImpl.validate(ValidatorHandlerImpl.java:730) ~[?:1.8.0_71]
	at com.sun.org.apache.xerces.internal.jaxp.validation.ValidatorImpl.validate(ValidatorImpl.java:102) ~[?:1.8.0_71]
	at javax.xml.validation.Validator.validate(Validator.java:124) ~[?:1.8.0_71]
	at com.hp.mqm.platform.security.InputValidationServiceImpl.validateSecureXML(InputValidationServiceImpl.java:401) ~[mqm-infra-services-impl-12.55.5.100.jar:?]
	at com.hp.mqm.testbox.TestResultPushServiceImpl.pushResults(TestResultPushServiceImpl.java:109) ~[test-box-service-12.55.5.100.jar:?]
	... 9 more
```

Here is my OctaneGherkinResults file, in the Jenkins server, at jobs/myjob/builds/8/OctaneGherkinResults0.xml:
```
<?xml version="1.0" encoding="UTF-8"?>
<features version="1">
  <feature name="Sottoscrizione di PIR con firma digitale" path="/var/jenkins_home/jobs/myjob/workspace/testsuite/target/test-classes/com/example/test/e2e/sottoscrizione.feature" started="1503408123523" tag="@TID1001REV0.2.0">
    <file><![CDATA[#Auto generated Octane revision tag
@TID1001REV0.2.0
Feature: Sottoscrizione

  blablah
  blablah

[..]

  Scenario: Sottometto form vuoto
    Given sono il family banker
    And propongo al cliente il contratto PIR in firma digitale
    When non compilo nessun campo
    And il cliente sottoscrive il contratto
    Then dovrebbe comparire un messaggio di errore

]]></file>
    <scenarios>

[..]

      <scenario name="Sottometto form vuoto">
        <steps>
          <step duration="369937550" name="Given sono il family banker" status="failed">
            <error_message><![CDATA[java.lang.RuntimeException: fake failure
	at com.example.test.e2e.steps.api.ApiStepdefs.givenCodiceAgente(ApiStepdefs.java:30)
	at ✽.Given sono il family banker 213993(com/example/test/e2e/sottoscrizione.feature:74)
]]></error_message>
          </step>
          <step duration="0" name="And propongo al cliente il contratto PIR in firma digitale" status="skipped"/>
          <step duration="0" name="When non compilo nessun campo"/>
          <step duration="0" name="And il cliente sottoscrive il contratto" status="skipped"/>
          <step duration="0" name="Then dovrebbe comparire un messaggio di errore"/>
        </steps>
      </scenario>
    </scenarios>
  </feature>
[..]
</features>

```

As you can see many steps in OctaneGherkinResults0.xml don't have the status attribute indeed.

It looks like that octane-cucumber-jvm is not generating properly the XML file.

I'm using:
- octane-cucumber-jvm  12.53.22,
- Octane 12.55.5.100 (on *docker*)

In the meantime, I'll try downgrading the version of Octane by using an older docker image.

Cheers

**Comment by alb-i986** [comment]
By debugging a test execution it's easy to see that the problem is in `HPEAlmOctaneGherkinFormatter` which doesn't properly set the status in all cases, see e.g. https://github.com/HPSoftware/octane-cucumber-jvm/blob/44e295362c0014e834ed64332b5442b0ef7220c1/src/main/java/com/hpe/alm/octane/HPEAlmOctaneGherkinFormatter.java#L164

**Comment by alb-i986** [comment]
As a workaround, users need to exclude from the execution any scenario with one or more pending steps.
One way of doing this is for example by tagging those scenarios with `@wip`, and setting the CucumberOptions like this:

```
@RunWith(OctaneCucumber.class)
@CucumberOptions(tags = {"~@wip"})
public class MyCukesTest {}
```

**Comment by sharonvardi** [comment]
@alb-i986 can you please provide us your step definition file and Test class so we can investigate this issue?

**Comment by alb-i986** [comment]
Sorry, I can't provide you with the exact content.
Anyway, it's not relevant. What matters is that one or more steps in a scenario must not be implemented.
As an example, please try the following:

Runner:
```
@RunWith(OctaneCucumber.class)
public class MyCukesTest {}
```

Stepdefs class:
```
public class Stepdefs {

    @Given("^I am logged in$")
    public void login() {
        // not relevant
    }
}
```

Feature file:
```
Feature: asd

  Scenario: scenario with pending steps
    Given I am logged in
    When pending step # this step is not implemented
    Then another pending step # this step is not implemented
```

When cucumber runs the second step ("When pending step"), what happens is that `HPEAlmOctaneGherkinFormatter#result` is called with an UNDEFINED Result object, therefore the status attribute *is not set*
https://github.com/HPSoftware/octane-cucumber-jvm/blob/44e295362c0014e834ed64332b5442b0ef7220c1/src/main/java/com/hpe/alm/octane/HPEAlmOctaneGherkinFormatter.java#L163

**Comment by sharonvardi** [comment]
The issue is fixed on version 12.55.7
